sdk/python/lib/pulumi/runtime/stack.py

```diff
@@ -29,49 +29,69 @@
     from .. import Output
 
 
+def _get_running_tasks() -> List[asyncio.Task]:
+    pending = []
+    for task in _all_tasks():
+        # Don't kill ourselves, that would be silly.
+        if not task == _get_current_task():
+            pending.append(task)
+    return pending
+
+
 async def run_pulumi_func(func: Callable):
     try:
         func()
     finally:
         log.debug("Waiting for outstanding RPCs to complete")
 
-        # Pump the event loop, giving all of the RPCs that we just queued up time to fully execute.
-        # The asyncio scheduler does not expose a "yield" primitive, so this will have to do.
-        #
-        # Note that "asyncio.sleep(0)" is the blessed way to do this:
-        # https://github.com/python/asyncio/issues/284#issuecomment-154180935
-        #
-        # We await each RPC in turn so that this loop will actually block rather than busy-wait.
         while True:
-            await asyncio.sleep(0)
-            rpcs_remaining = len(RPC_MANAGER.rpcs)
-            if rpcs_remaining == 0:
+            # Pump the event loop, giving all of the RPCs that we just queued up time to fully execute.
+            # The asyncio scheduler does not expose a "yield" primitive, so this will have to do.
+            #
+            # Note that "asyncio.sleep(0)" is the blessed way to do this:
+            # https://github.com/python/asyncio/issues/284#issuecomment-154180935
+            #
+            # We await each RPC in turn so that this loop will actually block rather than busy-wait.
+            while len(RPC_MANAGER.rpcs) > 0:
+                await asyncio.sleep(0)
+                log.debug(f"waiting for quiescence; {len(RPC_MANAGER.rpcs)} RPCs outstanding")
+                await RPC_MANAGER.rpcs.pop()
+
+            if RPC_MANAGER.unhandled_exception is not None:
+                raise RPC_MANAGER.unhandled_exception.with_traceback(RPC_MANAGER.exception_traceback)
+
+            log.debug("RPCs successfully completed")
+
+            # If the RPCs have successfully completed, now await all remaining outstanding tasks.
+            outstanding_tasks = _get_running_tasks()
+            if len(outstanding_tasks) == 0:
+                log.debug("No outstanding tasks to complete, exiting")
+            else:
+                log.debug(f"Waiting for {len(outstanding_tasks)} outstanding tasks to complete")
+
+                done, pending = await asyncio.wait(outstanding_tasks, return_when="FIRST_EXCEPTION")
```

Comment:
I'm a little bit uncertain about the behavior of `return_when="FIRST_EXCEPTION"`. From [the docs](https://docs.python.org/3.6/library/asyncio-task.html#asyncio.wait):

> The function will return when any future finishes by raising an exception. If no future raises an exception then it is equivalent to ALL_COMPLETED.

If this stops at the very first exception--i.e. before giving whatever is waiting on the task a chance to handle the exception--then I don't think that we should cancel pending tasks, and should instead use `ALL_COMPLETED`.

For example, consider the following:

```python

async def raises():
    await asyncio.sleep(0)
    raise Exception("oh no")

async def catches():
    try:
        await raises()
    except Exception:
        print("oh yeah")

catches()
```

If `return_when="FIRST_EXCEPTION"` returns after `raises` raises its exception without allowing `catches` a chance to run, then we'll cancel `catches` and unintentionally abort a program that would not fail.

Reply:
Ooh, excellent test case. This was good to check but `"FIRST_EXCEPTION"` actually does work the way I had thought it would in that it returns on the first "true" exception, i.e. one that isn't handled and actually bubbles through to the top-level task. I added the test case.